src/CoreWCF.Primitives/src/CoreWCF/Description/ServiceMetadataExtension.cs

```diff
@@ -87,40 +87,31 @@ public WsdlNS.ServiceDescription SingleWsdl
 
         internal Uri HttpsHelpPageUrl { get; set; }
 
-        internal bool UpdateAddressDynamically { get; set; }
+        internal bool UpdateAddressDynamically => DynamicMetadataEndpointAddressProvider != null;
 
         // This dictionary should not be mutated after open
         internal IDictionary<string, int> UpdatePortsByScheme { get; set; }
 
-        internal static bool TryGetHttpHostAndPort(Uri listenUri, HttpRequest httpRequest, out string host, out int port)
+        public IMetadataEndpointAddressProvider DynamicMetadataEndpointAddressProvider { get; internal set; }
+
+        internal bool TryGetHttpHostAndPort(Uri listenUri, HttpRequest httpRequest, out string host, out int port, out string scheme)
         {
             host = null;
             port = 0;
+            scheme = null;
 
-            // Get the host header
-            HostString hostString = httpRequest.Host;
-            if (!hostString.HasValue)
-            {
-                return false;
-            }
-
-            host = hostString.Host;
-            if (hostString.Port.HasValue)
-            {
-                port = hostString.Port.Value;
-            }
-            else
+            Uri customUri = DynamicMetadataEndpointAddressProvider?.GetEndpointAddress(httpRequest);
+            if (customUri != null)
             {
```

Comment:
This is a change in behavior. If DynamicMetadataEndpointAddressProvider isn't set (the default with no behaviors), we are now returning false. Previously it would directly pull from the request and if there was a Host header, it would return true. Now it returns false if an address provider isn't set.

Reply:
good catch. Reviewing carefully I also noticed that I may have changed behavior because previously  we were using the `_listenUri` to deduct the port using `Uri.TryCreate`. Now i no longer have access to the listenUri in the `MetadataEndpointAddressProvider` and i do not think we want to polluate the signature of `GetEndpointAddress` with it. I don't know what is the best way to get it in the `GetEndpointAddress` implementation.

Reply:
> This is a change in behavior. If DynamicMetadataEndpointAddressProvider isn't set (the default with no behaviors), we are now returning false. Previously it would directly pull from the request and if there was a Host header, it would return true. Now it returns false if an address provider isn't set.

I just checked and `TryGetHttpHostAndPort` is always called from `GetDynamicAddressWriter` calls which are always surrounded by a `if (UpdateAddressDynamically)` which was previously only set when `UseRequestHeadersForMetadataAddressBehavior` was applied. We can then deduct TryGetHttpHostAndPort is executed only when an address provider in the new code given that UseRequestHeadersForMetadataAddressBehavior  also provides an address provider.

Reply:
I just worked that out (the if check) trying to go through the change very carefully. HttpContext has an [Items property bag](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpcontext.items?view=aspnetcore-6.0) you can put things in. You could stash the listenUri in there. I generally use the convention of the key being a string with the fully qualified class name of the either the object (which doesn't apply here as Uri is generic), or the owner or interested party of the object. That way it's impossible to have an accidental key name collision. In this case the value comes from `CoreWCF.Description.ServiceMetadataExtension.HttpGetImpl._listenUri` and is initially passed into a method in `TryHandleMetadataRequestAsync`. You could put the listenUri in the Items property bag with that fully qualified field name that I just mentioned.